### DISCUSSION TITLE
how to support FreePad_LeTV_8（zigbee）

### DISCUSSION THREAD

**Discussion by qlwz** [arendst/Tasmota#10399] [Q&A] [answered]
@s-hadinger 

the z2m config 
`    {
        zigbeeModel: ['FreePad_LeTV_8'],
        model: 'FreePad_LeTV_8',
        vendor: 'DIYRuZ',
        description: '[LeTV 8key FreePad mod](https://modkam.ru/?p=1791)',
        fromZigbee: [fz.diyruz_freepad_clicks, fz.diyruz_freepad_config, fz.battery],
        exposes: [e.battery(), e.action(['*_single', '*_double', '*_triple', '*_quadruple', '*_release'])].concat(((enpoinsCount) => {
            const features = [];
            for (let i = 1; i <= enpoinsCount; i++) {
                const epName = `button_${i}`;
                features.push(
                    exposes.enum('switch_type', exposes.access.ALL, ['toggle', 'momentary', 'multifunction']).withEndpoint(epName));
                features.push(exposes.enum('switch_actions', exposes.access.ALL, ['on', 'off', 'toggle']).withEndpoint(epName));
            }
            return features;
        })(8)),
        toZigbee: [tz.diyruz_freepad_on_off_config, tz.factory_reset],
        meta: {configureKey: 1},
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint = device.getEndpoint(1);
            await reporting.bind(endpoint, coordinatorEndpoint, ['genPowerCfg']);
            if (device.applicationVersion < 3) { // Legacy PM2 firmwares
                const payload = [{
                    attribute: 'batteryPercentageRemaining', minimumReportInterval: 0, maximumReportInterval: 3600, reportableChange: 0,
                }, {
                    attribute: 'batteryVoltage', minimumReportInterval: 0, maximumReportInterval: 3600, reportableChange: 0,
                }];
                await endpoint.configureReporting('genPowerCfg', payload);
            }
            device.endpoints.forEach(async (ep) => {
                if (ep.outputClusters.includes(18)) {
                    await reporting.bind(ep, coordinatorEndpoint, ['genMultistateInput']);
                }
            });
        },
        endpoint: (device) => {
            return {button_1: 1, button_2: 2, button_3: 3, button_4: 4, button_5: 5, button_6: 6, button_7: 7, button_8: 8};
        },
    },`

how to send 
`            device.endpoints.forEach(async (ep) => {
                if (ep.outputClusters.includes(18)) {
                    await reporting.bind(ep, coordinatorEndpoint, ['genMultistateInput']);
                }
            });
`?

**Comment by s-hadinger** [comment]
It should work out of the box, but it will probably be of lower level than the transformation of events included in Z2M.

Do you have the device? Can you send logs using `WebLog 3`?

**Comment by qlwz** [comment]
00:00:00.037 CFG: Loaded, 数量: 9
00:00:00.163 QPC: Count 1
00:00:00.169 ZIG: Randomizing Zigbee parameters, please check with 'ZbConfig'
00:00:00.173 项目: tasmota Tasmota 版本 9.2.0.2(tasmota)-1_0_5_RC4(2021-01-05T15:02:58)
00:00:00.971 WIF: 连接到 AP1 zeki Channel 10 BSSId D8:C8:E9:E1:63:80 模式: 11n 名称: tasmota_21E15C-0348...
00:00:04.870 WIF: 已连接
00:00:05.126 HTP: Web 服务器地址: tasmota_21E15C-0348 IP 地址: 192.168.1.212
08:05:24.075 RSL: tele/tasmota_21E15C/INFO1 = {"Module":"ESP32-DevKit","Version":"9.2.0.2(tasmota)","FallbackTopic":"cmnd/DVES_21E15C_fb/","GroupTopic":"cmnd/tasmotas/"}
08:05:24.080 RSL: tele/tasmota_21E15C/INFO2 = {"WebServerMode":"Admin","Hostname":"tasmota_21E15C-0348","IPAddress":"192.168.1.212"}
08:05:24.092 RSL: tele/tasmota_21E15C/INFO3 = {"RestartReason":"Software reset CPU"}
08:05:24.782 QPC: Reset
08:05:30.109 RSL: tele/tasmota_21E15C/STATE = {"Time":"2021-01-05T08:05:30","Uptime":"0T00:00:14","UptimeSec":14,"Heap":111,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":264,"MqttCount":0,"Wifi":{"AP":1,"SSId":"zeki","BSSId":"D8:C8:E9:E1:63:80","Channel":10,"RSSI":100,"Signal":-42,"LinkCount":1,"Downtime":"0T00:00:06"}}
08:05:33.547 ZIG: rebooting CC2530 device
08:05:34.800 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":1,"Message":"CC2530 booted","RestartReason":"External","MajorRel":2,"MinorRel":6}}
08:05:34.000 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":50,"MajorRel":2,"MinorRel":6,"MaintRel":3,"Revision":20190619}}
08:05:35.407 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":3,"Message":"Configured, starting coordinator"}}
08:05:36.256 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":40,"NewState":9,"Message":"Started as coordinator"}}
08:05:36.307 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":51,"IEEEAddr":"0x00124B001EC874F9","ShortAddr":"0x0000","DeviceType":7,"DeviceState":9,"NumAssocDevices":2,"AssocDevicesList":["0x2021","0xDEB6"]}}
08:05:36.856 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":0,"Message":"Started"}}
08:05:36.857 ZIG: Zigbee started
08:05:36.858 TFS: Mounting...
E (20463) SPIFFS: mount failed, -10025
08:05:36.863 TFS: Formatting...
08:05:37.812 TFS: Mounted
08:05:37.826 TFS: File not found
08:05:37.827 ZIG: No Zigbee device information in Flash
08:06:17.404 CMD: WebLog 3
08:06:17.409 RSL: stat/tasmota_21E15C/RESULT = {"WebLog":3}
08:06:51.917 RSL: stat/tasmota_21E15C/RESULT = {"ZbPermitJoin":"Done"}
08:06:51.991 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":21,"Message":"Enable Pairing
mode for 60 seconds"}}
08:07:19.467 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":34,"IEEEAddr":"0x00124B000ADB81F7","ShortAddr":"0xDEB6","ParentNetwork":"0x0000"}}
08:07:21.117 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":30,"IEEEAddr":"0x00124B000ADB81F7","ShortAddr":"0xDEB6","PowerSource":false,"ReceiveWhenIdle":false,"Security":false}}
08:07:21.417 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":32,"ActiveEndpoints":["0x08","0x07","0x06","0x05","0x04","0x03","0x02","0x01"]}}
08:07:21.525 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xDEB6":{"Device":"0xDEB6","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":80}}}
08:07:21.544 TFS: File not found
08:07:21.654 ZIG: Zigbee Devices Data saved in Flash (32 bytes)
08:07:21.668 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xDEB6":{"Device":"0xDEB6","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":80}}}
08:07:23.811 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":33,"Device":"0xDEB6","Endpoint":"0x08","ProfileId":"0x0104","DeviceId":"0x0006","DeviceVersion":2,"InClusters":[],"OutClusters":["0x0006","0x0012"]}}
08:07:24.962 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":33,"Device":"0xDEB6","Endpoint":"0x07","ProfileId":"0x0104","DeviceId":"0x0006","DeviceVersion":2,"InClusters":[],"OutClusters":["0x0006","0x0012"]}}
08:07:26.112 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":33,"Device":"0xDEB6","Endpoint":"0x06","ProfileId":"0x0104","DeviceId":"0x0006","DeviceVersion":2,"InClusters":[],"OutClusters":["0x0006","0x0012"]}}
08:07:28.263 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":33,"Device":"0xDEB6","Endpoint":"0x05","ProfileId":"0x0104","DeviceId":"0x0006","DeviceVersion":2,"InClusters":[],"OutClusters":["0x0006","0x0012","0x0008","0x0300"]}}
08:07:29.458 ZIG: auto-bind `ZbBind {"Device":"0xDEB6","Endpoint":8,"Cluster":"0x0006"}`
08:07:29.761 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xDEB6","Status":0,"StatusMessage":"SUCCESS"}}
08:07:31.458 ZIG: auto-bind `ZbBind {"Device":"0xDEB6","Endpoint":7,"Cluster":"0x0006"}`
08:07:33.458 ZIG: auto-bind `ZbBind {"Device":"0xDEB6","Endpoint":6,"Cluster":"0x0006"}`
08:07:35.458 ZIG: auto-bind `ZbBind {"Device":"0xDEB6","Endpoint":5,"Cluster":"0x0006"}`
08:07:36.517 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xDEB6":{"Device":"0xDEB6","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":96}}}
08:07:36.567 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xDEB6":{"Device":"0xDEB6","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":96}}}
08:07:36.661 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xDEB6","Status":0,"StatusMessage":"SUCCESS"}}
08:07:37.110 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xDEB6","Status":0,"StatusMessage":"SUCCESS"}}
08:07:37.458 ZIG: auto-bind `ZbBind {"Device":"0xDEB6","Endpoint":5,"Cluster":"0x0008"}`
08:07:37.760 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xDEB6","Status":0,"StatusMessage":"SUCCESS"}}
08:07:38.261 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xDEB6","Status":0,"StatusMessage":"SUCCESS"}}
08:07:39.458 ZIG: auto-bind `ZbBind {"Device":"0xDEB6","Endpoint":5,"Cluster":"0x0300"}`
08:07:51.362 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":34,"IEEEAddr":"0x00124B000ADB81F7","ShortAddr":"0xA471","ParentNetwork":"0x0000"}}
08:07:52.712 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":30,"IEEEAddr":"0x00124B000ADB81F7","ShortAddr":"0xA471","PowerSource":false,"ReceiveWhenIdle":false,"Security":false}}
08:07:52.918 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":57}}}
08:07:52.968 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":57}}}
08:07:53.063 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":32,"ActiveEndpoints":["0x08","0x07","0x06","0x05","0x04","0x03","0x02","0x01"]}}
08:07:53.531 ZIG: Zigbee Devices Data saved in Flash (32 bytes)
08:07:56.640 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":33,"Device":"0xA471","Endpoint":"0x07","ProfileId":"0x0104","DeviceId":"0x0006","DeviceVersion":2,"InClusters":[],"OutClusters":["0x0006","0x0012"]}}
08:07:57.740 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":33,"Device":"0xA471","Endpoint":"0x06","ProfileId":"0x0104","DeviceId":"0x0006","DeviceVersion":2,"InClusters":[],"OutClusters":["0x0006","0x0012"]}}
08:07:59.891 RSL: tele/tasmota_21E15C/RESULT = {"ZbState":{"Status":33,"Device":"0xA471","Endpoint":"0x05","ProfileId":"0x0104","DeviceId":"0x0006","DeviceVersion":2,"InClusters":[],"OutClusters":["0x0006","0x0012","0x0008","0x0300"]}}
08:08:01.087 ZIG: auto-bind `ZbBind {"Device":"0xA471","Endpoint":7,"Cluster":"0x0006"}`
08:08:02.040 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xA471","Status":0,"StatusMessage":"SUCCESS"}}
08:08:03.087 ZIG: auto-bind `ZbBind {"Device":"0xA471","Endpoint":6,"Cluster":"0x0006"}`
08:08:05.087 ZIG: auto-bind `ZbBind {"Device":"0xA471","Endpoint":5,"Cluster":"0x0006"}`
08:08:07.086 ZIG: auto-bind `ZbBind {"Device":"0xA471","Endpoint":5,"Cluster":"0x0008"}`
08:08:07.946 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":98}}}
08:08:07.996 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":98}}}
08:08:08.040 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xA471","Status":0,"StatusMessage":"SUCCESS"}}
08:08:08.740 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xA471","Status":0,"StatusMessage":"SUCCESS"}}
08:08:09.040 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xA471","Status":0,"StatusMessage":"SUCCESS"}}
08:08:09.087 ZIG: auto-bind `ZbBind {"Device":"0xA471","Endpoint":5,"Cluster":"0x0300"}`
08:08:09.190 RSL: tele/tasmota_21E15C/RESULT = {"ZbBind":{"Device":"0xA471","Status":0,"StatusMessage":"SUCCESS"}}
08:08:22.947 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":88}}}
08:08:22.996 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":88}}}
08:08:37.943 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":70}}}
08:08:37.993 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":70}}}
08:08:48.286 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0006!02":"","Power":2,"Endpoint":6,"LinkQuality":46}}}
08:08:48.336 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0008!06":"002B0A00","DimmerUp":true,"Endpoint":5,"LinkQuality":49}}}
08:08:48.388 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0300!09":"E80300000000","ColorStep":[1000,0],"Endpoint":5,"LinkQuality":52}}}
08:08:48.438 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0300!09":"18FCE8030000","ColorStep":[64536,1000],"Endpoint":5,"LinkQuality":52}}}
08:08:51.984 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0300!09":"18FCE8030000","ColorStep":[64536,1000],"Endpoint":5,"LinkQuality":62}}}
08:08:52.932 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":70}}}
08:08:52.981 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":70}}}
08:09:22.955 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":28}}}
08:09:22.973 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":28}}}
08:09:37.957 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":67}}}
08:09:38.006 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":67}}}
08:09:52.955 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":78}}}
08:09:53.004 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":78}}}
08:10:07.913 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":67}}}
08:10:07.962 RSL: tele/tasmota_21E15C/SENSOR = {"ZbReceived":{"0xA471":{"Device":"0xA471","0003!01":"","IdentifyQuery":true,"Endpoint":8,"LinkQuality":67}}}


https://github.com/diyruz/freepad  this is firmware url

**Comment by qlwz** [comment]
[log.txt](https://github.com/arendst/Tasmota/files/5768637/log.txt)

**Comment by qlwz** [comment]
@s-hadinger 
{"ZbState":{"Status":32,"ActiveEndpoints":["0x08","0x07","0x06","0x05","0x04","0x03","0x02","0x01"]}}

but only bind  Endpoint：8 7 6 5

edit Z_ReceiveActiveEp ->  if ((i < 4) && (ep < 0x10)) {   to  if (/*(i < 4) && */(ep < 0x10)) {   
now bind 8 7 6 5 4   but  no bind 3 2 1

**Comment by s-hadinger** [comment, marked as answer]
Indeed, I limited auto-binding to the first 4 endpoints. Can you send me the logs at `Weblog 3` when you removed the limit of 4?

**Comment by sztsian** [comment]
Hi,

I'm also trying to use LeTV 8 key remote controller with Free Pad. My tasmota setup is a ESP32Solo with EZSP (IKEA TRADFRI with [this firmware](https://github.com/MattWestb/IKEA-TRADFRI-ICC-A-1-Module)). The tasmota firmware is compiled by TasmoCompiler for 9.5.0 with only Zigbee & EZSP configured by user override. Like the comment on Jan 6th, only 4 button works, which are just the button 5-8 on [this article](https://diyruz.github.io/posts/letv/). Attached are the logs of pressing button 8,7,6,5 with `weblog 3`.

[LeTV_FreePad_press_tasmota32.txt](https://github.com/arendst/Tasmota/files/7022468/LeTV_FreePad_press_tasmota32.txt)

So my questions are

- Is there a way to use the other 4 buttons
- When pressing button 6 (I call it `Microphone button` in the log), it shows event with both endpoint 6 and endpoint 5. Is it some issue of the FreePad firmware, or it's probably a bug somewhere in Tasmota?
- I want to config this remote in either Home Assistant or directly in Tasmota by rules. Shall I directly use the following message `MQT: tele/tasmota_2398E8/SENSOR = {"ZbReceived":{"0x26E8":{"Device":"0x26E8","0300!09":"18FCE8030000","ColorStep":[64536,1000],"Endpoint":5,"LinkQuality":92}}}` and use Endpoint 5 as the condition, or there is a even better message to use as condition?

Unfortunately I wasn't able to figure out where is the limit of 4 set in the code (as @s-hadinger  mentioned in the comment of Jan 9) so I cannot provide any such info from my side. But if anyone can provide me a compiled tasmota firmware with the 4 endpoint limit revoked, I'd like to help testing and provide the log of pressing such buttons.

**Reply by s-hadinger** [reply]
The limit of 4 auto-binding can be raised but I doubt that the device will stay awake long enough for all the commands to succeed. Please share the full logs at loglevel 3 when you pair the device?

**Reply by s-hadinger** [reply]
The limit of 4 is here
https://github.com/arendst/Tasmota/blob/e5e8ef47366f6a49da88533f8c35433e78a259fd/tasmota/xdrv_23_zigbee_8_parsers.ino#L641

**Reply by sztsian** [reply]
The pairing log is available here
[LeTV_FreePad_pair_tasmota32.txt](https://github.com/arendst/Tasmota/files/7024676/LeTV_FreePad_pair_tasmota32.txt)

**Reply by sztsian** [reply]
I don't understand the code there but I decided to test by changing from `(i < 4)` to `(i < 0)` (to just disable the function block) and `(i < 8)` (in hope of all the 8 activeEpCount can be parsed in the function block). If I change to `(i < 0)`, pairing works  just fine but the remote seems totally not working, which is (probably) expected? And then I change to `(i < 8)`, looks like things are much better, each button responses with some mqtt message

So now I'm also attaching the whole log (in `weblog 3`) with `(i < 8)` code. It also contains the log when I press all the buttons (in random order)
[LeTV_FreePad_pairpress_if-i-lt-8_tasmota32.txt](https://github.com/arendst/Tasmota/files/7024689/LeTV_FreePad_pairpress_if-i-lt-8_tasmota32.txt)

**Reply by sztsian** [reply]
Oh by the way, I'm changing on top of the 9.5.0 code tarball not the git head.

**Reply by s-hadinger** [reply]
Thanks. Happy that it finally works for you. I will change the limit to 8 in the code. I had doubts but you confirmed that it was useful.
